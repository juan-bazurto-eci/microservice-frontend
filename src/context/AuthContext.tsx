import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  ReactNode,
} from "react";
import { AuthContextType, User } from "../types/authTypes";
import { useRouter } from "next/router";

const AuthContext = createContext<AuthContextType | undefined>(undefined);

interface AuthProviderProps {
  children: ReactNode;
}

export const AuthProvider = ({ children }: AuthProviderProps) => {
  const [user, setUser] = useState<User | null>(null);
  const router = useRouter();
  useEffect(() => {
    const userData = localStorage.getItem("userData");
    if (userData) {
      setUser(JSON.parse(userData));
    }
  }, []);

  const login = (userData: User) => {
    localStorage.setItem("userData", JSON.stringify(userData));
    setUser(userData);
  };

  const logout = () => {
    localStorage.removeItem("userData");
    setUser(null);
    router.push("/");
  };

  const isUserLoggedIn = () => user !== null;

  const getUserData = () => user;

  const updateUser = (updatedUserData: User) => {
    localStorage.setItem("userData", JSON.stringify(updatedUserData));
    setUser(updatedUserData);
  };

  return (
    <AuthContext.Provider
      value={{ user, login, logout, isUserLoggedIn, getUserData, updateUser }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = (): AuthContextType => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
