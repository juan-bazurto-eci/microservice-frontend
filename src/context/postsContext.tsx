import { PostType } from "@/types/postType";
import React, { createContext, useContext, useState } from "react";
import { getMaxPostId } from "@/utils/getMaxPostId";

export type PostContextType = {
  lengthPost: number;
  posts: PostType[] | never[];
  addPost: (post: PostType) => void;
  setPosts: (posts: PostType[]) => void;
  updatePost: (updatedPost: PostType) => void;
  removePost: (id: number) => void;
  searchPost: (id: number) => PostType | undefined;
};

const PostsContext = createContext<PostContextType>({
  lengthPost: 0,
  posts: [],
  addPost: (post) => {},
  updatePost: (updatedPost) => {},
  setPosts: () => {},
  removePost: (id) => {},
  searchPost: (id) => {
    return {};
  },
});

export const PostsProvider = ({ children }: any) => {
  const [posts, setPostsProvider] = useState<PostType[] | never[]>([]);

  const setPosts = (posts: PostType[]) => {
    setPostsProvider(posts);
    localStorage.setItem("posts", JSON.stringify(posts));
  };

  const addPost = (post: PostType) => {
    setPostsProvider((prevPosts) => [...prevPosts, post]);
    localStorage.setItem("posts", JSON.stringify([...posts, post]));
  };

  const removePost = (id: number) => {
    setPostsProvider((prevPosts) =>
      prevPosts.filter((post) => post?.id !== id)
    );
    localStorage.setItem(
      "posts",
      JSON.stringify([...posts.filter((post) => post?.id !== id)])
    );
  };

  const searchPost = (id: number) => {
    return posts.find((post) => post?.id === id);
  };

  const updatePost = (updatedPost: PostType) => {
    const updatedIndex = posts.findIndex((post) => post.id === updatedPost.id);
    if (updatedIndex !== -1) {
      setPostsProvider((prevPosts) => [
        ...prevPosts.slice(0, updatedIndex),
        { ...prevPosts[updatedIndex], ...updatedPost },
        ...prevPosts.slice(updatedIndex + 1),
      ]);
      localStorage.setItem(
        "posts",
        JSON.stringify([
          ...posts.slice(0, updatedIndex),
          { ...posts[updatedIndex], ...updatedPost },
          ...posts.slice(updatedIndex + 1),
        ])
      );
    }
  };

  return (
    <PostsContext.Provider
      value={{
        lengthPost: getMaxPostId(posts) ?? 0,
        posts,
        addPost,
        setPosts,
        removePost,
        searchPost,
        updatePost,
      }}
    >
      {children}
    </PostsContext.Provider>
  );
};

export const usePostsContext = () => {
  return useContext(PostsContext);
};
