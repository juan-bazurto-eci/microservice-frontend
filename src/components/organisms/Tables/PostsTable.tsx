import React, { useEffect } from "react";
import { useTheme } from "@mui/material/styles";
import {
  Box,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TablePagination,
  TableRow,
  TableSortLabel,
  Toolbar,
  FormControlLabel,
  Typography,
  Paper,
} from "@mui/material";
import { visuallyHidden } from "@mui/utils";
import CustomCheckbox from "@/components/atoms/checkbox/CustomCheckbox";
import { PostType } from "@/types/postType";
import CustomSwitch from "@/components/atoms/switch/CustomSwitch";
import EnhancedTableToolbar from "@/components/molecules/toolbar/EnhancedTableToolbar";
import AlertSubmmit from "@/components/atoms/alert/AlertSubmmit";
import axios from "axios";
import { ConsultPostDTO } from "@/models/ConsultPostDTO";
import { usePostsContext } from "@/context/postsContext";
import { isValidArray } from "@/utils/isValidArray";

function descendingComparator<T>(a: T, b: T, orderBy: keyof T) {
  if (b[orderBy] < a[orderBy]) {
    return -1;
  }
  if (b[orderBy] > a[orderBy]) {
    return 1;
  }

  return 0;
}

type Order = "asc" | "desc";

function getComparator<Key extends keyof any>(
  order: Order,
  orderBy: Key
): (
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  a: { [key in Key]: number | string },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  b: { [key in Key]: number | string }
) => number {
  return order === "desc"
    ? (a, b) => descendingComparator(a, b, orderBy)
    : (a, b) => -descendingComparator(a, b, orderBy);
}

function stableSort<T>(array: T[], comparator: (a: T, b: T) => number) {
  const stabilizedThis = array.map((el, index) => [el, index] as [T, number]);
  stabilizedThis.sort((a, b) => {
    const order = comparator(a[0], b[0]);
    if (order !== 0) {
      return order;
    }

    return a[1] - b[1];
  });

  return stabilizedThis.map((el) => el[0]);
}

interface HeadCell {
  disablePadding: boolean;
  id: string;
  label: string;
  numeric: boolean;
}

const headCells: readonly HeadCell[] = [
  {
    id: "id",
    numeric: true,
    disablePadding: false,
    label: "ID",
  },
  {
    id: "title",
    numeric: false,
    disablePadding: false,
    label: "Título",
  },

  {
    id: "body",
    numeric: false,
    disablePadding: false,
    label: "Descripción",
  },
];

interface EnhancedTableProps {
  numSelected: number;
  onRequestSort: (event: React.MouseEvent<unknown>, property: any) => void;
  onSelectAllClick: (event: React.ChangeEvent<HTMLInputElement>) => void;
  order: Order;
  orderBy: string;
  rowCount: number;
}

function EnhancedTableHead(props: EnhancedTableProps) {
  const {
    onSelectAllClick,
    order,
    orderBy,
    numSelected,
    rowCount,
    onRequestSort,
  } = props;
  const createSortHandler =
    (property: any) => (event: React.MouseEvent<unknown>) => {
      onRequestSort(event, property);
    };

  return (
    <TableHead>
      <TableRow>
        <TableCell padding="checkbox">
          <CustomCheckbox
            color="primary"
            checked={rowCount > 0 && numSelected === rowCount}
            onChange={onSelectAllClick}
            inputProps={{
              "aria-label": "select all rows",
            }}
          />
        </TableCell>
        {headCells.map((headCell) => (
          <TableCell
            key={headCell.id}
            align={headCell.numeric ? "right" : "left"}
            padding={headCell.disablePadding ? "none" : "normal"}
            sortDirection={orderBy === headCell.id ? order : false}
          >
            <TableSortLabel
              active={orderBy === headCell.id}
              direction={orderBy === headCell.id ? order : "asc"}
              onClick={createSortHandler(headCell.id)}
            >
              {headCell.label}
              {orderBy === headCell.id ? (
                <Box component="span" sx={visuallyHidden}>
                  {order === "desc" ? "sorted descending" : "sorted ascending"}
                </Box>
              ) : null}
            </TableSortLabel>
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  );
}

const PostsTable = ({ dashboard = false }: Props) => {
  const { posts, setPosts, removePost } = usePostsContext();
  const [order, setOrder] = React.useState<Order>("asc");
  const [orderBy, setOrderBy] = React.useState<any>("title");
  const [selected, setSelected] = React.useState<readonly string[]>([]);
  const [page, setPage] = React.useState(0);
  const [dense, setDense] = React.useState(false);
  const [deleteSuccess, setDeleteSuccess] = React.useState(false);
  const [rowsPerPage, setRowsPerPage] = React.useState(10);
  const [rows, setRows] = React.useState<any>(posts);
  const [search, setSearch] = React.useState("");

  const fetchData = async () => {
    try {
      const consultedPosts =
        JSON.parse(localStorage.getItem("posts") || "[]") ?? [];
      if (!isValidArray(consultedPosts)) {
        const response = await axios.get(`/api/posts/get`);
        setPosts(ConsultPostDTO(response?.data ?? {}));
      } else {
        setPosts(ConsultPostDTO(consultedPosts ?? {}));
      }
    } catch (error) {}
  };

  useEffect(() => {
    if (!isValidArray(posts)) {
      fetchData();
    }
  }, []);

  const handleSearch = (event: React.ChangeEvent<HTMLInputElement>) => {
    const filteredRows: PostType[] = posts?.filter((row: PostType) => {
      return row?.title
        ?.toString()
        ?.toLowerCase()
        ?.includes(event.target.value);
    });
    setSearch(event.target.value);
    setRows(filteredRows);
    setPage(0);
  };

  // This is for the sorting
  const handleRequestSort = (
    event: React.MouseEvent<unknown>,
    property: any
  ) => {
    const isAsc = orderBy === property && order === "asc";
    setOrder(isAsc ? "desc" : "asc");
    setOrderBy(property);
  };

  // This is for select all the row
  const handleSelectAllClick = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.checked) {
      const newSelecteds = rows.map((n: any) => n.id);
      setSelected(newSelecteds);

      return;
    }
    setSelected([]);
  };

  // This is for the single row sleect
  const handleClick = (event: React.MouseEvent<unknown>, name: string) => {
    const selectedIndex = selected.indexOf(name);
    let newSelected: readonly string[] = [];

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, name);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1)
      );
    }

    setSelected(newSelected);
  };

  const handleCloseAlert = () => {
    setDeleteSuccess(false);
  };

  const handleDeleteAlert = (status: number) => {
    setSelected([]);
    setDeleteSuccess(true);
  };

  const handleDelete = async () => {
    try {
      let successResponse = 404;
      for (const id of selected) {
        const response = await axios
          .delete(`/api/posts/delete/${id}`)
          ?.then((response) => {
            return response;
          });
        if (response?.status === 200) {
          successResponse = 200;
          removePost(parseInt(id));
        }
      }
      handleDeleteAlert(successResponse);
    } catch (error) {
      console.error(error);
    }
  };

  const handleChangePage = (event: unknown, newPage: number) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setRowsPerPage(parseInt(event.target.value, 10));
    setPage(0);
  };

  const handleChangeDense = (event: React.ChangeEvent<HTMLInputElement>) => {
    setDense(event.target.checked);
  };

  const isSelected = (name: string) => selected.indexOf(name) !== -1;

  const emptyRows =
    page > 0 ? Math.max(0, (1 + page) * rowsPerPage - rows.length) : 0;

  const theme = useTheme();
  const borderColor = theme.palette.divider;

  React.useEffect(() => {
    setRows(posts);
  }, [posts]);

  return (
    <>
      <Box>
        <Box>
          {dashboard ? (
            <EnhancedTableToolbar
              numSelected={selected.length}
              search={search}
              handleSearch={(event: any) => handleSearch(event)}
              handleDelete={handleDelete}
            />
          ) : null}
          <Paper
            variant="outlined"
            sx={{ mx: 2, mt: 1, border: `1px solid ${borderColor}` }}
          >
            <TableContainer>
              <Table
                sx={{ minWidth: 750 }}
                aria-labelledby="tableTitle"
                size={dense ? "small" : "medium"}
              >
                <EnhancedTableHead
                  numSelected={selected.length}
                  order={order}
                  orderBy={orderBy}
                  onSelectAllClick={handleSelectAllClick}
                  onRequestSort={handleRequestSort}
                  rowCount={rows.length}
                />
                <TableBody>
                  {stableSort(rows, getComparator(order, orderBy))
                    .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                    .map((row: any, index) => {
                      const isItemSelected = isSelected(row?.id);
                      const labelId = `enhanced-table-checkbox-${index}`;
                      return (
                        <TableRow
                          hover
                          onClick={(event) => handleClick(event, row?.id)}
                          role="checkbox"
                          aria-checked={isItemSelected}
                          tabIndex={-1}
                          key={`${row?.userId}-${row?.id}-${index}`}
                          selected={isItemSelected}
                        >
                          <TableCell padding="checkbox">
                            <CustomCheckbox
                              color="primary"
                              checked={isItemSelected}
                              inputProps={{
                                "aria-labelledby": labelId,
                              }}
                            />
                          </TableCell>

                          <TableCell>
                            <Typography variant="h6" fontWeight="400">
                              {row?.id ?? ""}
                            </Typography>
                          </TableCell>
                          <TableCell>
                            <Typography
                              color="textSecondary"
                              variant="subtitle2"
                              fontWeight="400"
                            >
                              {row?.title ?? ""}
                            </Typography>
                          </TableCell>
                          <TableCell>
                            <Typography fontWeight={600} variant="h6">
                              {row?.body ?? ""}
                            </Typography>
                          </TableCell>
                        </TableRow>
                      );
                    })}
                  {emptyRows > 0 && (
                    <TableRow
                      style={{
                        height: (dense ? 33 : 53) * emptyRows,
                      }}
                    >
                      <TableCell colSpan={6} />
                    </TableRow>
                  )}
                </TableBody>
              </Table>
            </TableContainer>
            <TablePagination
              rowsPerPageOptions={[5, 10, 25]}
              component="div"
              count={rows.length}
              rowsPerPage={rowsPerPage}
              page={page}
              onPageChange={handleChangePage}
              onRowsPerPageChange={handleChangeRowsPerPage}
            />
          </Paper>
          <Toolbar
            sx={{
              pl: { sm: 2 },
              pr: { xs: 1, sm: 1 },
            }}
          >
            <Box sx={{ flex: "1 1 100%" }}>
              <FormControlLabel
                control={
                  <CustomSwitch checked={dense} onChange={handleChangeDense} />
                }
                label="Dense padding"
              />
            </Box>
          </Toolbar>
        </Box>
      </Box>
      <AlertSubmmit
        open={deleteSuccess}
        handleClose={handleCloseAlert}
        title={"Posts eliminados correctamente!"}
        severity={"success"}
      />
    </>
  );
};

interface Props {
  posts?: PostType[];
  dashboard?: boolean;
}

export default PostsTable;
