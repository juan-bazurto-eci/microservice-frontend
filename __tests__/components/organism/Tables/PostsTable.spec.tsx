import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { PostsProvider } from "@/context/postsContext";
import PostsTable from "@/components/organisms/Tables/PostsTable";

describe("PostsTable Component", () => {
  const theme = createTheme({
    breakpoints: {
      values: {
        xs: 0,
        sm: 600,
        md: 960,
        lg: 1280,
        xl: 1920,
      },
    },
  });

  const posts = [
    { id: 1, title: "Title 1", body: "Body 1" },
    { id: 2, title: "Title 2", body: "Body 2" },
    { id: 3, title: "Title 3", body: "Body 3" },
  ];

  it("renders PostsTable", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <PostsProvider posts={posts}>
          <PostsTable />
        </PostsProvider>
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("sorts rows when clicking on column headers", () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <PostsProvider posts={posts}>
          <PostsTable />
        </PostsProvider>
      </ThemeProvider>
    );

    fireEvent.click(getByText("TÃ­tulo"));

    fireEvent.click(getByText("ID"));
  });

  it("selects rows when clicking on checkboxes", () => {
    const { getByLabelText } = render(
      <ThemeProvider theme={theme}>
        <PostsProvider posts={posts}>
          <PostsTable />
        </PostsProvider>
      </ThemeProvider>
    );

    fireEvent.click(getByLabelText("select all rows"));
  });

  it("handles page change correctly", () => {
    const { getByLabelText } = render(
      <ThemeProvider theme={theme}>
        <PostsProvider posts={posts}>
          <PostsTable />
        </PostsProvider>
      </ThemeProvider>
    );

    fireEvent.click(getByLabelText("Go to next page"));
  });

  it("handles dense padding switch correctly", () => {
    const { getByLabelText } = render(
      <ThemeProvider theme={theme}>
        <PostsProvider posts={posts}>
          <PostsTable />
        </PostsProvider>
      </ThemeProvider>
    );

    fireEvent.click(getByLabelText("Dense padding"));
  });
});
