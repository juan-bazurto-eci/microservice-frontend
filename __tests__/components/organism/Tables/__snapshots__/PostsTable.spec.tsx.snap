// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsTable Component renders PostsTable 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-1ddowl6-MuiPaper-root"
      >
        <div
          class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
        >
          <table
            aria-labelledby="tableTitle"
            class="MuiTable-root css-1gplja-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingCheckbox MuiTableCell-sizeMedium css-hbtnrh-MuiTableCell-root"
                  scope="col"
                >
                  <span
                    class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-355xbt-MuiButtonBase-root-MuiCheckbox-root"
                  >
                    <input
                      aria-label="select all rows"
                      class="PrivateSwitchBase-input css-1m9pwf3"
                      data-indeterminate="false"
                      type="checkbox"
                    />
                    <span
                      class="css-1j4k40g"
                      font-size="medium"
                    />
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                  scope="col"
                >
                  <span
                    class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                    role="button"
                    tabindex="0"
                  >
                    ID
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                      data-testid="ArrowDownwardIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
                <th
                  aria-sort="ascending"
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  <span
                    class="MuiButtonBase-root MuiTableSortLabel-root Mui-active css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                    role="button"
                    tabindex="0"
                  >
                    Título
                    <span
                      class="MuiBox-root css-trfuzi"
                    >
                      sorted ascending
                    </span>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                      data-testid="ArrowDownwardIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  <span
                    class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                    role="button"
                    tabindex="0"
                  >
                    Descripción
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                      data-testid="ArrowDownwardIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
          </table>
        </div>
        <div
          class="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
          >
            <div
              class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
            />
            <p
              class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
              id=":r1:"
            >
              Rows per page:
            </p>
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-16c50h-MuiInputBase-root-MuiTablePagination-select"
            >
              <div
                aria-controls=":r2:"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby=":r1: :r0:"
                class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
                id=":r0:"
                role="combobox"
                tabindex="0"
              >
                10
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="10"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
            >
              0–0 of 0
            </p>
            <div
              class="MuiTablePagination-actions"
            >
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                tabindex="-1"
                title="Go to previous page"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowLeftIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                  />
                </svg>
              </button>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                tabindex="-1"
                title="Go to next page"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-pjjt8i-MuiToolbar-root"
      >
        <div
          class="MuiBox-root css-1mrwq1p"
        >
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiSwitch-root MuiSwitch-sizeMedium css-1ddo3z7-MuiSwitch-root"
            >
              <span
                class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
              >
                <input
                  class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                  type="checkbox"
                />
                <span
                  class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              Dense padding
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
